File: src/main/java/com/lightbend/beam/ModelServer.java
package com.lightbend.beam;

import com.lightbend.kafka.JobConfiguration;
import com.lightbend.kafka.KafkaOptions;
import com.lightbend.model.Model;
import com.lightbend.model.Winerecord;
import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.coders.ByteArrayCoder;
import org.apache.beam.sdk.coders.KvCoder;
import org.apache.beam.sdk.coders.SerializableCoder;
import org.apache.beam.sdk.io.kafka.KafkaIO;
import org.apache.beam.sdk.transforms.*;
import org.apache.beam.sdk.transforms.windowing.AfterWatermark;
import org.apache.beam.sdk.transforms.windowing.FixedWindows;
import org.apache.beam.sdk.transforms.windowing.Window;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollection;
import org.apache.beam.sdk.values.PCollectionView;
import org.joda.time.Duration;

import java.util.Arrays;
import java.util.Iterator;

/**
 * Created by boris on 5/17/17.
 * Loosely based on
 *  https://beam.apache.org/blog/2017/02/13/stateful-processing.html
 */
public class ModelServer {

    public static void main(String[] args) {

        // Create and initialize pipeline
        Pipeline p = JobConfiguration.initializePipeline(args);
        KafkaOptions options = p.getOptions().as(KafkaOptions.class);

        KvCoder<byte[], byte[]> kafkaDataCoder = KvCoder.of(ByteArrayCoder.of(), ByteArrayCoder.of());
        SerializableCoder<Model> modelsCoder = SerializableCoder.of(Model.class);

        // Data Stream
/*        PCollection<Winerecord.WineRecord> dataStream = p
                .apply("data", KafkaIO.readBytes()
                        .withBootstrapServers(options.getBroker())
                        .withTopics(Arrays.asList(options.getKafkaDataTopic()))
                        .updateConsumerProperties(JobConfiguration.getKafkaConsumerProps(options, true))
                        .withoutMetadata()).setCoder(kafkaDataCoder)
                // Group into fixed-size windows
                .apply(Window.<KV<byte[], byte[]>>into(FixedWindows.of(
                        Duration.standardSeconds(1)))
                        .triggering(AfterWatermark.pastEndOfWindow()).
                                withAllowedLateness(Duration.ZERO)
                        .discardingFiredPanes())
                .apply("Parse data records", ParDo.of(new DataRecordProcessor.ConvertDataRecordFunction()));*/

        PCollection<KV<String,Iterable<Winerecord.WineRecord>>> dataStream = p
                .apply("data", KafkaIO.readBytes()
                        .withBootstrapServers(options.getBroker())
                        .withTopics(Arrays.asList(options.getKafkaDataTopic()))
                        .updateConsumerProperties(JobConfiguration.getKafkaConsumerProps(options, true))
                        .withoutMetadata()).setCoder(kafkaDataCoder)
                // Group into fixed-size windows
                .apply(Window.<KV<byte[], byte[]>>into(FixedWindows.of(
                        Duration.standardSeconds(1)))
                        .triggering(AfterWatermark.pastEndOfWindow()).
                                withAllowedLateness(Duration.ZERO)
                        .discardingFiredPanes())
                .apply("Parse data records", ParDo.of(new DataRecordProcessor.ConvertDataRecordToKVFunction()))
                .apply("GroupByKey",GroupByKey.<String, Winerecord.WineRecord>create());

        // Models Stream
        PCollection<ModelRecordProcessor.ModelToServe> modelStream = p
                .apply("models", KafkaIO.readBytes()
                        .withBootstrapServers(options.getBroker())
                        .withTopics(Arrays.asList(options.getKafkaModelsTopic()))
                        .updateConsumerProperties(JobConfiguration.getKafkaConsumerProps(options, false))
                        .withoutMetadata()).setCoder(kafkaDataCoder)
                // Group into fixed-size windows
                .apply(Window.<KV<byte[], byte[]>>into(FixedWindows.of(
                        Duration.standardSeconds(1)))
                        .triggering(AfterWatermark.pastEndOfWindow()).
                                withAllowedLateness(Duration.ZERO)
                        .discardingFiredPanes())
                .apply("Parse model", ParDo.of(new ModelRecordProcessor.ConvertModelRecordFunction()))
                .apply("Print ", MapElements.via(new DataRecordProcessor.PrinterSimpleFn<ModelRecordProcessor.ModelToServe>()));

        // Build a model view to use as a side input
        PCollectionView<Model> models = modelStream
                .apply("building model state",
                        Combine.globally(new ModelRecordProcessor.ModelFromEventsFn()).withoutDefaults().asSingletonView());

        PCollection<Double> scoringResults = dataStream
                .apply("Scoring the model", ParDo.withSideInputs(models).of(
                        new DoFn<KV<String,Iterable<Winerecord.WineRecord>>, Double>(){

                    @ProcessElement
                    public void processElement(ProcessContext ctx){
                        KV<String,Iterable<Winerecord.WineRecord>> record = ctx.element();
                        Model model = ctx.sideInput(models);
                        if(model == null)
                            System.out.println("No model available - skipping");
                        else{
                            // Ignore key for now
                            Iterator<Winerecord.WineRecord> iterator = record.getValue().iterator();
                            while(iterator.hasNext()) {
                                long start = System.currentTimeMillis();
                                double quality = (double) model.score(iterator.next());
                                long duration = System.currentTimeMillis() - start;
                                System.out.println("Calculated quality - " + quality + " in " + duration + "ms");
                                ctx.output(quality);
                            }
                        }
                    };
                }));
        p.run();
    }
}

File: server/src/main/scala/com/lightbend/scala/beam/ModelServer.scala
package com.lightbend.scala.beam

import java.util

import com.lightbend.scala.beam.processors.{ConvertDataRecordFn, ConvertModelRecordFn, ScoringFn, SimplePrinterFn}
import com.lightbend.kafka.JobConfiguration
import com.lightbend.modelServer.model.ModelWithData
import com.lightbend.scala.beam.coders.ScalaDoubleCoder
import org.apache.beam.sdk.Pipeline
import org.apache.beam.sdk.coders.{ByteArrayCoder, KvCoder, NullableCoder}
import org.apache.beam.sdk.io.kafka.KafkaIO
import org.apache.beam.sdk.transforms.{Flatten, MapElements, ParDo}
import org.apache.beam.sdk.values.{KV, PCollectionList}

object ModelServer {

  def main(args: Array[String]): Unit = {
    // Create and initialize pipeline
    val options = JobConfiguration.initializePipeline(args)
    val p = Pipeline.create(options)

    // Coder to use for Kafka data - raw byte message
    val kafkaDataCoder = KvCoder.of(NullableCoder.of(ByteArrayCoder.of), ByteArrayCoder.of)

    // Data Stream - gets data records from Kafka topic
    // It has to use KV to work with state, which is always KV
    val dataStream = p
        .apply("data", KafkaIO.readBytes
          .withBootstrapServers(options.getBroker)
          .withTopics(util.Arrays.asList(options.getKafkaDataTopic))
          .updateConsumerProperties(JobConfiguration.getKafkaConsumerProps(options, true))
          .withoutMetadata).setCoder(kafkaDataCoder)
        // Convert Kafka message to WineRecord
        .apply("Parse data records", ParDo.of(new ConvertDataRecordFn))

    // Models Stream - get model records from Kafka
    // It has to use KV to work with state, which is always KV
    val modelStream = p
          .apply("models", KafkaIO.readBytes
            .withBootstrapServers(options.getBroker)
            .withTopics(util.Arrays.asList(options.getKafkaModelsTopic))
            .updateConsumerProperties(JobConfiguration.getKafkaConsumerProps(options, false))
            .withoutMetadata).setCoder(kafkaDataCoder)
          // Convert Kafka record to ModelDescriptor
          .apply("Parse model", ParDo.of(new ConvertModelRecordFn))

    // Create a combined PCollection stream combining both model and data streams.
    val combinedStream = PCollectionList.of(dataStream).and(modelStream)
      // Flatten the list
      .apply(Flatten.pCollections[KV[String, ModelWithData]])

    // Score models and print result
    val scoringResults = combinedStream
      // Score data using current model
      .apply("Scoring the model", ParDo.of(new ScoringFn)).setCoder(ScalaDoubleCoder.of)
      // Print scoring result
      .apply("Print result", MapElements.via(new SimplePrinterFn[Double]))

    // Run the pipeline
    p.run
  }
}